stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/socket/OpenSecureSocket.java

```diff
@@ -0,0 +1,206 @@
+import org.ballerinalang.natives.annotations.ReturnType;
+import org.ballerinalang.util.codegen.PackageInfo;
+import org.ballerinalang.util.codegen.StructInfo;
+import org.ballerinalang.util.exceptions.BallerinaException;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.Socket;
+import java.nio.channels.ByteChannel;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Security;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+/**
+ * Native function to open a secure client socket.
+ *
+ * @since 0.964.0
+ */
+@BallerinaFunction(packageName = "ballerina.io",
+                   functionName = "openSecureSocket",
+                   args = {
+                           @Argument(name = "host", type = TypeKind.STRING),
+                           @Argument(name = "port", type = TypeKind.INT),
+                           @Argument(name = "option", type = TypeKind.STRUCT, structType = "SocketProperties",
+                                     structPackage = "ballerina.io")
+                   },
+                   returnType = {
+                           @ReturnType(type = TypeKind.STRUCT,
+                                       structType = "Socket",
+                                       structPackage = "ballerina.io")
+                   },
+                   isPublic = true)
+public class OpenSecureSocket extends AbstractNativeFunction {
+
+    private static final String SEPARATOR = ",";
+    private static final String SOCKET_PACKAGE = "ballerina.io";
+    private static final String SOCKET_STRUCT_TYPE = "Socket";
+    private static final String BYTE_CHANNEL_STRUCT_TYPE = "ByteChannel";
+
+    @Override
+    public BValue[] execute(Context context) {
+        final String host = getStringArgument(context, 0);
+        final int port = (int) getIntArgument(context, 0);
+        final BStruct options = (BStruct) getRefArgument(context, 0);
+        Socket socket;
+        try {
+            SSLContext sslContext = getSslContext(options);
+            SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+            socket = sslSocketFactory.createSocket(host, port);
+        } catch (Throwable e) {
+            String msg = "Failed to open a connection to [" + host + ":" + port + "] : " + e.getMessage();
+            throw new BallerinaException(msg, e, context);
+        }
+        SSLSocket sslSocket;
+        ByteChannel channel;
+        try {
+            if (socket instanceof SSLSocket) {
```

Comment:
any possibility of getting all these lines into a single try/catch block ?

Reply:
This will help to identify the keystore/truststore/handshake errors and connectivity error separately. We will again refactor this once the IOError struct available.